src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs

```diff
@@ -39,6 +39,38 @@ public CertificateCallbackMapper(Func<HttpRequestMessage, X509Certificate2, X509
         }
 
         public static async ValueTask<(Socket, Stream)> ConnectAsync(string host, int port, CancellationToken cancellationToken)
+        {
+            CancellationTokenSource cancelIPv6 = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+            Task<(Socket, Stream)> tryConnectAsyncIPv6 = ConnectAsyncInternal(host, port, AddressFamily.InterNetworkV6, cancelIPv6.Token);
+
+            if (await Task.WhenAny(tryConnectAsyncIPv6, Task.Delay(200)).ConfigureAwait(false) == tryConnectAsyncIPv6 && tryConnectAsyncIPv6.IsCompletedSuccessfully)
+            {
+                return await tryConnectAsyncIPv6.ConfigureAwait(false);
+            }
+
+            CancellationTokenSource cancelIPv4 = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+            Task<(Socket, Stream)> tryConnectAsyncIPv4 = ConnectAsyncInternal(host, port, AddressFamily.InterNetwork, cancelIPv4.Token);
+            if (await Task.WhenAny(tryConnectAsyncIPv6, tryConnectAsyncIPv4).ConfigureAwait(false) == tryConnectAsyncIPv6)
+            {
+                if (tryConnectAsyncIPv6.IsCompletedSuccessfully)
+                {
+                    cancelIPv4.Cancel();
+                    return await tryConnectAsyncIPv6.ConfigureAwait(false);
+                }
+                return await tryConnectAsyncIPv4.ConfigureAwait(false);
+            }
+            else
+            {
+                if (tryConnectAsyncIPv4.IsCompletedSuccessfully)
+                {
+                    cancelIPv6.Cancel();
+                    return await tryConnectAsyncIPv4.ConfigureAwait(false);
+                }
+                return await tryConnectAsyncIPv6.ConfigureAwait(false);
+            }
```

Comment:
Seems like you could avoid some duplication here:
```C#
Task<(Socket, Stream)> completedTask = await Task.WhenAny(tryConnectAsyncIPv6, tryConnectAsyncIPv4).ConfigureAwait(false);
Task<(Socket, Stream)> otherTask = completedTask == tryConnectAsyncIPv6 ? tryConnectAsyncIPv4 : tryConnectAsyncIPv6;
if (completedTask.IsCompletedSuccessfully)
{
    cancelIP.Cancel();
    return completedTask.GetAwaiter().GetResult();
}
return await otherTask.ConfigureAwait(false);
```

Reply:
Combining the cancellation tokens and rewriting makes this way cleaner. Thanks!